codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,68 +242,28 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
-            }
         } catch (Http2NoMoreStreamIdsException e) {
             lifecycleManager.onException(ctx, e);
             return promise.setFailure(e);
         } catch (Throwable e) {
             return promise.setFailure(e);
         }
-
-        if (lastDataWrite == null) {
-            // No previous DATA frames to keep in sync with, just send it now.
-            return writeHeaders(ctx, stream, headers, streamDependency, weight, exclusive, padding,
-                    endOfStream, promise);
-        }
-
-        // There were previous DATA frames sent.  We need to send the HEADERS only after the most
-        // recent DATA frame to keep them in sync...
-
-        // Only write the HEADERS frame after the previous DATA frame has been written.
-        final Http2Stream theStream = stream;
-        lastDataWrite.addListener(new ChannelFutureListener() {
-            @Override
-            public void operationComplete(ChannelFuture future) throws Exception {
-                if (!future.isSuccess()) {
-                    // The DATA write failed, also fail this write.
-                    promise.setFailure(future.cause());
-                    return;
-                }
-
-                // Perform the write.
-                writeHeaders(ctx, theStream, headers, streamDependency, weight, exclusive, padding,
-                        endOfStream, promise);
+        ChannelFuture future = flowController().sendHeaders(ctx, stream, headers, streamDependency, weight,
+            exclusive, padding, endOfStream, promise);
+        future.addListener(new ChannelFutureListener() {
+          @Override
+          public void operationComplete(ChannelFuture future) throws Exception {
+            if (!future.isSuccess()) {
+              // The write failed, handle the error.
+              lifecycleManager.onException(ctx, future.cause());
+            } else if (endOfStream) {
+              lifecycleManager.closeLocalSide(stream, ctx.newPromise());
```

Comment:
Good catch. Ill make it synchronous here so that no further writes can be enqueued.